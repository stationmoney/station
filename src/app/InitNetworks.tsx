import { PropsWithChildren, useEffect, useState } from "react"
import axios from "axios"
import { ASSETS } from "config/constants"
import createContext from "utils/createContext"
import NetworkLoading from "./NetworkLoading"

export const [useNetworks, NetworksProvider] = createContext<{
  networks: InterchainNetworks
  filterEnabledNetworks: <T>(network: Record<string, T>) => Record<string, T>
}>("useNetworks")

const InitNetworks = ({ children }: PropsWithChildren<{}>) => {
  const [networks, setNetworks] = useState<InterchainNetworks>()
  const [enabledNetworks, setEnabledNetworks] = useState<string[]>([])

  useEffect(() => {
    const fetchChains = async () => {
      const { data: chains } = await axios.get<InterchainNetworks>(
        "/station/chains.json",
        {
          baseURL: ASSETS,
        }
      )
      setNetworks(chains)
    }

    fetchChains()
  }, [])

  useEffect(() => {
    const testChains = async () => {
      if (!networks) return
      const testBase = {
        ...networks.mainnet,
        ...networks.testnet,
        ...networks.classic,
      }

      const result = await Promise.all(
        Object.values(testBase).map(async (network) => {
          try {
            const { data } = await axios.get(
              "cosmos/base/tendermint/v1beta1/node_info",
              {
                baseURL: network.lcd,
                timeout: 5_000,
              }
            )
            return data.default_node_info.network as string
          } catch (e) {
            return null
          }
        })
      )

      setEnabledNetworks(
        result.filter((r) => typeof r === "string") as string[]
      )
    }

    testChains()
  }, [networks])

  if (!networks || !enabledNetworks.length)
    return <NetworkLoading title="Connecting to available networks..." />

  return (
    <NetworksProvider
      value={{
        networks,
        filterEnabledNetworks: (networks) =>
          Object.fromEntries(
            Object.entries(networks).filter(([chainID]) =>
              enabledNetworks.includes(chainID)
            )
          ),
      }}
    >
      {children}
    </NetworksProvider>
  )
}

export default InitNetworks
