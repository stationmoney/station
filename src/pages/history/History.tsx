import { useTranslation } from "react-i18next"
import { useIsTerraAPIAvailable } from "data/Terra/TerraAPI"
import { Wrong } from "components/feedback"
import HistoryList from "./HistoryList"
import { ChainFilter, Page } from "components/layout"

const History = () => {
  const { t } = useTranslation()
  const available = useIsTerraAPIAvailable()

  if (!available) return <Wrong>{t("History is not supported")}</Wrong>

  return (
    <Page title={t("History")}>
      <ChainFilter outside>
        {(chain) => <HistoryList chainID={chain ?? ""} />}
      </ChainFilter>
    </Page>
  )
}

export default History
