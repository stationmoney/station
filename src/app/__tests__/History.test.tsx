import React from "react"
import History from "../../pages/history/History"
import { render } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import { RecoilRoot } from "recoil"
import { NetworksProvider } from "../InitNetworks"
import { useQuery } from "react-query"
import { mockHistory } from "./__mocks__/History.mock"
import { mockNetworks } from "./__mocks__/Networks.mock"
import { mockWhitelist } from "./__mocks__/Whitelist.mock"
import { mockInterchainAddresses } from "./__mocks__/InterchainAddresses.mock"

function renderComponent() {
  type TokenFilter = <T>(network: Record<string, T>) => Record<string, T>
  const networks = {} as jest.Mocked<InterchainNetworks>
  const mockedTokenFilter = {} as jest.Mocked<TokenFilter>

  return render(
    <NetworksProvider
      value={{
        networks: networks,
        filterEnabledNetworks: mockedTokenFilter,
        filterDisabledNetworks: mockedTokenFilter,
      }}
    >
      <MemoryRouter>
        <RecoilRoot>
          <History />
        </RecoilRoot>
      </MemoryRouter>
    </NetworksProvider>
  )
}

jest.mock("../../auth/hooks/useAddress", () => {
  const useInterchainAddresses = () => {
    return mockInterchainAddresses
  }

  return {
    useInterchainAddresses: useInterchainAddresses,
  }
})

jest.mock("../../data/wallet", () => {
  const mockUseAddress = () => {
    return "terra111111111111111111111111111111111111111"
  }

  const mockUseNetwork = () => {
    return mockNetworks
  }

  const mockUseNetworkName = () => {
    return "testnet"
  }

  return {
    useAddress: mockUseAddress,
    useNetwork: mockUseNetwork,
    useNetworkName: mockUseNetworkName,
  }
})

jest.mock("../../data/queries/chains", () => {
  const actual = jest.requireActual("../../data/queries/chains")

  const mockUseWhitelist = () => {
    return mockWhitelist
  }

  return {
    ...actual,
    useWhitelist: mockUseWhitelist,
  }
})

jest.mock("react-query", () => ({
  useQuery: jest.fn().mockReturnValue({
    data: Object,
    isLoading: Boolean,
    error: Object,
  }),
}))

describe("History component (containing HistoryList/Item/Message subcomponents) matches snapshots", () => {
  it("matches original component", () => {
    useQuery.mockReturnValue({
      data: mockHistory,
      isLoading: false,
      error: {},
    })
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })
})
