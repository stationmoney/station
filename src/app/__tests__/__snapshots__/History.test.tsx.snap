// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History component (containing HistoryList/Item/Message subcomponents) matches snapshots matches original component 1`] = `
<DocumentFragment>
  <article
    class="page"
  >
    <div
      class="container grid"
    >
      <header
        class="header"
      >
        <div
          class="titleWrapper"
        >
          <h1
            class="title"
          >
            History
          </h1>
        </div>
      </header>
      <section
        class="main"
      >
        <div
          class="chainfilter__out"
        >
          <div
            class="header"
          >
            <div
              class="pills"
            >
              <button
                class="active"
              >
                <img
                  alt="Terra"
                  src="https://station-assets.terra.money/img/chains/Terra.svg"
                />
                Terra
              </button>
              <button>
                <img
                  alt="Atreides"
                  src="https://station-assets.terra.money/img/chains/Atreides.png"
                />
                Atreides
              </button>
              <button>
                <img
                  alt="Corrino"
                  src="https://station-assets.terra.money/img/chains/Corrino.png"
                />
                Corrino
              </button>
              <button>
                <img
                  alt="Harkonnen"
                  src="https://station-assets.terra.money/img/chains/Harkonnen.png"
                />
                Harkonnen
              </button>
              <button>
                <img
                  alt="Ordos"
                  src="https://station-assets.terra.money/img/chains/Ordos.png"
                />
                Ordos
              </button>
            </div>
          </div>
          <div
            class="content"
          >
            <article
              class="page invisible"
            >
              <div
                class="container grid"
              >
                <section
                  class="main"
                >
                  <div
                    class="col"
                    style="flex: 1;"
                  >
                    <article
                      class="card small bordered"
                    >
                      <section
                        class="main"
                      >
                        <header
                          class="header"
                        >
                          <p
                            class="txhash"
                          >
                            <span
                              class="chain"
                            >
                              <img
                                alt="pisco-1"
                                src="https://station-assets.terra.money/img/chains/Terra.svg"
                              />
                              Terra
                            </span>
                            <span
                              class="link"
                            >
                              <a
                                class="link"
                                href="https://terrascope.info/testnet/tx/1126F13EAD4ECBC590478A6C055271297044A8E57B6908E2844D7B964D064E67"
                                rel="noopener noreferrer"
                                target="_blank"
                              >
                                1126F1...064E67
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                  data-testid="CallMadeIcon"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M9 5v2h6.59L4 18.59 5.41 20 17 8.41V15h2V5z"
                                  />
                                </svg>
                              </a>
                            </span>
                          </p>
                          <p
                            class="time"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                              data-testid="DateRangeIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M9 11H7v2h2v-2zm4 0h-2v2h2v-2zm4 0h-2v2h2v-2zm2-7h-1V2h-2v2H8V2H6v2H5c-1.11 0-1.99.9-1.99 2L3 20c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 16H5V9h14v11z"
                              />
                            </svg>
                            <span>
                              15 days ago
                            </span>
                          </p>
                        </header>
                        <div
                          class="msgs"
                        >
                          <div
                            class="component"
                          >
                            <span
                              class="inline tag bg-info"
                            >
                              Send
                            </span>
                            <section>
                              <p
                                class="message"
                              >
                                <span>
                                  <a
                                    class="link"
                                    href="https://terrascope.info/testnet/address/terra18j6kg55a960v0l845mxeg6k0r8vfux5tsexdkp"
                                    rel="noopener noreferrer"
                                    target="_blank"
                                    value="terra18j6kg55a960v0l845mxeg6k0r8vfux5tsexdkp"
                                  >
                                    my wallet
                                    <svg
                                      aria-hidden="true"
                                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                      data-testid="CallMadeIcon"
                                      focusable="false"
                                      viewBox="0 0 24 24"
                                    >
                                      <path
                                        d="M9 5v2h6.59L4 18.59 5.41 20 17 8.41V15h2V5z"
                                      />
                                    </svg>
                                  </a>
                                </span>
                                <span>
                                   send
                                </span>
                                <span>
                                   
                                  <span
                                    class="textmain"
                                  >
                                    <span
                                      class="component"
                                    >
                                      1,000
                                      <span
                                        class="small"
                                      >
                                        .00
                                      </span>
                                      <span
                                        class="small"
                                      >
                                         ORD
                                      </span>
                                    </span>
                                  </span>
                                </span>
                                <span>
                                   to
                                </span>
                                <span>
                                   
                                  <a
                                    class="link"
                                    href="https://terrascope.info/testnet/address/terra1mfqnlqz5hh7wzy2lcfmw82s5aqtx2l9efafqqr"
                                    rel="noopener noreferrer"
                                    target="_blank"
                                    value="terra1mfqnlqz5hh7wzy2lcfmw82s5aqtx2l9efafqqr"
                                  >
                                    terra1...fafqqr
                                    <svg
                                      aria-hidden="true"
                                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                      data-testid="CallMadeIcon"
                                      focusable="false"
                                      viewBox="0 0 24 24"
                                    >
                                      <path
                                        d="M9 5v2h6.59L4 18.59 5.41 20 17 8.41V15h2V5z"
                                      />
                                    </svg>
                                  </a>
                                </span>
                              </p>
                            </section>
                          </div>
                        </div>
                        <footer
                          class="footer"
                        >
                          <dl
                            class="dl"
                          >
                            <dt>
                              Fee
                            </dt>
                            <dd>
                              <span
                                class="component"
                              >
                                0
                                <span
                                  class="small"
                                >
                                  .003432
                                </span>
                                <span
                                  class="small"
                                >
                                   LUNA
                                </span>
                              </span>
                            </dd>
                          </dl>
                        </footer>
                      </section>
                    </article>
                  </div>
                </section>
              </div>
            </article>
          </div>
        </div>
      </section>
    </div>
  </article>
</DocumentFragment>
`;
